File: backend/app/services/ai_text.py
"""
AI 文本处理工具（Phase 3.6 AI Copilot）

集中维护对 Dify / LLM 返回文本的清洗与归一化逻辑，避免业务代码各自重复正则。

设计要点：
- 纯函数，无外部依赖，便于单元测试
- 所有函数接受 ``text: str | None`` 入参，None 时返回空串（前端展示安全）
- 输出与业务语义解耦：剥离噪声但不修改业务内容
"""
from __future__ import annotations

import re
from typing import Optional


# ---------------------------------------------------------------------------
# 剥离模型推理过程块
# ---------------------------------------------------------------------------
# 推理类模型（DeepSeek R1、o1、o3-mini 等）在 Chat App response_mode=blocking
# 下，可能把推理过程以 ``...`` 的形式混入 ``answer`` 字段。
# 终态用户不需要看到这些，剥离并保留正文。
# - DOTALL：跨越换行
# - 非贪婪：避免跨多个 `` 块合并
_THINK_BLOCK_RE = re.compile(r"\<think\>.*?\</think\>", re.DOTALL)
# 折叠剥离后留下的连续空行（剥离 `` 常在 ``... `` 之间留下一片空行）
_BLANK_LINES_RE = re.compile(r"\n{3,}")


def strip_think_blocks(text: Optional[str]) -> str:
    """剥离文本中的 `` 推理过程块。

    Args:
        text: 原始文本（可能为 None）

    Returns:
        剥离后的纯答案文本（首尾空白已清理，连续空行已折叠为最多 2 个）

    Examples:
        >>> strip_think_blocks("\<think\>我先思考\<\/think\>你好")
        '你好'
        >>> strip_think_blocks(None)
        ''
        >>> strip_think_blocks("\<think\>思考\<\/think\>\\n\\n\\n正式答案")
        '正式答案'
    """
    if not text:
        return ""
    no_think = _THINK_BLOCK_RE.sub("", text)
    collapsed = _BLANK_LINES_RE.sub("\n\n", no_think)
    return collapsed.strip()


def has_think_block(text: Optional[str]) -> bool:
    """判断文本是否包含 `` 块（用于日志 / 监控埋点）。"""
    if not text:
        return False
    return _THINK_BLOCK_RE.search(text) is not None

File: frontend/src/utils/aiText.ts
/**
 * AI 文本处理工具（Phase 3.6 AI Copilot — refactor: hide ``）
 *
 * 与后端 `backend/app/services/ai_text.py` 行为一致：
 * - 推理模型（DeepSeek R1 / o1 / o3-mini 等）可能把 ``...`` 混入可见 answer
 * - 兜底剥离，避免历史脏数据 / 其它端点漏剥离时泄露给用户
 *
 * 设计要点：
 * - 纯函数，零依赖
 * - 空值安全（Null / Undefined → ''）
 * - 与 backend strip_think_blocks 行为完全对齐（DOTALL + 非贪婪 + 折叠空行）
 */

const THINK_BLOCK_RE = /\<think\>[\s\S]*?\<\/think\>/g
const BLANK_LINES_RE = /\n{3,}/g

/**
 * 剥离 AI 文本中的 `` 推理过程块。
 *
 * @example
 *   stripThinkBlocks('\<think\>我先思考\<\/think\>你好') // → '你好'
 *   stripThinkBlocks(null)                              // → ''
 */
export function stripThinkBlocks(text: string | null | undefined): string {
  if (!text) return ''
  const noThink = text.replace(THINK_BLOCK_RE, '')
  const collapsed = noThink.replace(BLANK_LINES_RE, '\n\n')
  return collapsed.trim()
}

/** 判断文本是否包含 `` 块（用于埋点 / 调试）。 */
export function hasThinkBlock(text: string | null | undefined): boolean {
  if (!text) return false
  return THINK_BLOCK_RE.test(text)
}
